### PR TITLE
fix agent thought bug and update the iot tool and the agent prompt

### PR DESCRIPTION
**Pull request by ruanrongman**
## Summary by CodeRabbit

- **New Features**
	- Enhanced agent interaction by introducing a `thought` parameter to improve action handling.
- **Improvements**
	- Updated IoT agent usage script for better clarity in agent identity, goals, and constraints.
	- Refined user instructions and control flow in the IoT switch MQTT prompt for enhanced user experience.

### CODE REVIEW

**Comment by Undertone0809** [comment]
Can you optimize revelant unit test to assert it?

https://github.com/Undertone0809/promptulate/blob/main/tests/hook/test_agent.py

**Comment by Undertone0809** [comment]
Moreover, there are some problem in `/test/hook/test_agent.py`. When we `agent.run()` https://github.com/Undertone0809/promptulate/blob/main/tests/hook/test_agent.py#L76, wo should inject mock data instread of  using true LLM request to get answer.

File: example/agent/iot_agent_usage.py [hunk 1]
```diff
@@ -41,9 +41,12 @@ def main():
         tools,
         enable_role=True,
         agent_name="小创",
```
Comment (Undertone0809) [inline example/agent/iot_agent_usage.py:43]: English name is better

File: example/agent/iot_agent_usage.py [hunk 2]
```diff
@@ -41,9 +41,12 @@ def main():
...
-        agent_goal="控制智能家居，可以开冷气，加热器以及开灯，完成动作后进入闲聊模式",
-        agent_constraints="在控制开关电器之前请尽量询问人类",
+        agent_identity="Smart speaker.",
+        agent_goal="Control smart home, can turn on the "
+        "air conditioner, heater, and lights, and enter into chat mode "
+        "after completing the action.",
+        agent_constraints="Please try to ask humans before controlling "
+        "or switching on electrical appliances.",
     )
     prompt = """我现在感觉好暗。"""
```
Comment (Undertone0809) [inline example/agent/iot_agent_usage.py:51]: English is better

File: promptulate/agents/tool_agent/agent.py [hunk 3]
```diff
@@ -127,11 +127,12 @@ def _run(self, prompt: str, *args, **kwargs) -> str:
             if "Final Answer" in answer:
                 return answer.split("Final Answer:")[-1]
 
-            action, action_input = self._find_action(answer)
+            action, thought, action_input = self._find_action(answer)
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:130]: thought, action, action_input is a better return value.

Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:130]: May be is function should be change to `_parse_llm_response(answer)`.

File: promptulate/agents/tool_agent/agent.py [hunk 4]
```diff
@@ -127,11 +127,12 @@ def _run(self, prompt: str, *args, **kwargs) -> str:
...
             Hook.call_hook(
                 HookTable.ON_AGENT_ACTION,
                 self,
                 action=action,
+                thought=thought,
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:135]: action and action_input two parameters should be adjacent.

File: promptulate/agents/tool_agent/agent.py [hunk 5]
```diff
@@ -167,25 +168,30 @@ def _find_action(self, answer: str) -> (str, str):
             answer(str): output of LLM
 
         Returns:
-            Return a tuple, (action, action input)
+            Return a tuple, (action,thought, action input)
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:171]: Change the revelant docs by upper modify.

File: promptulate/agents/tool_agent/agent.py [hunk 6]
```diff
@@ -167,25 +168,30 @@ def _find_action(self, answer: str) -> (str, str):
...
             action(str): tool name
             action_input(str): tool parameters
         """
         action_pattern = r"Action:\s*([\w-]+)"
+        thought_pattern = r"Thought:\s*([\w-]+)"
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:176]: action and action_input two parameters should be adjacent.

File: promptulate/agents/tool_agent/agent.py [hunk 7]
```diff
@@ -167,25 +168,30 @@ def _find_action(self, answer: str) -> (str, str):
...
         action_input_match = re.search(action_input_pattern, answer)
         action: str = ""
+        thought: str = ""
         action_input: str = ""
 
         if action_match:
             action = action_match.group(1)
             logger.info(f"[pne] tool agent get Action <{action}>")
-
+        if thought_match:
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:188]: action and action_input two parameters should be adjacent.

File: example/agent/iot_agent_usage.py (hunk 1 above)
Comment (ruanrongman) [reply example/agent/iot_agent_usage.py:43]: ok

File: example/agent/iot_agent_usage.py (hunk 2 above)
Comment (ruanrongman) [reply example/agent/iot_agent_usage.py:51]: ok

File: promptulate/agents/tool_agent/agent.py (hunk 3 above)
Comment (ruanrongman) [reply promptulate/agents/tool_agent/agent.py:130]: I think so.

File: tests/hook/test_agent.py [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 from promptulate.agents import ToolAgent
 from promptulate.hook import Hook
+from promptulate.llms import ErnieBot
```
Comment (Undertone0809) [inline tests/hook/test_agent.py:5]: Remove it

File: tests/hook/test_agent.py [hunk 9]
```diff
@@ -25,6 +27,21 @@ def handle_start(*args, **kwargs):
...
+            nonlocal action_flag
+            action_flag = True
+
+            thought = kwargs["thought"]
+            action = kwargs["action"]
+            action_input = kwargs["action_input"]
+            self.assertIsNotNone(thought)
+            self.assertIsNotNone(action)
+            self.assertIsNotNone(action_input)
+            print(f"<instance> thought: {thought}")
```
Comment (Undertone0809) [inline tests/hook/test_agent.py:41]: assert related string is better, rather than only not None

File: tests/hook/test_agent.py [hunk 10]
```diff
@@ -62,6 +81,21 @@ def handle_start(*args, **kwargs):
...
             print(f"<component> input: {inputs}")
 
+        @Hook.on_agent_action(hook_type="component")
+        def handle_action(*args, **kwargs):
+            nonlocal action_flag
+            action_flag = True
+
+            thought = kwargs["thought"]
+            action = kwargs["action"]
+            action_input = kwargs["action_input"]
```
Comment (Undertone0809) [inline tests/hook/test_agent.py:91]: if kwargs["thought"] is None, this line will get error

Comment (Undertone0809) [reply tests/hook/test_agent.py:91]: Optimize the test code please

File: promptulate/agents/tool_agent/agent.py [hunk 11]
```diff
@@ -160,32 +161,37 @@ def _should_continue(self, current_iteration: int, current_time_elapsed) -> bool
...
-    def _find_action(self, answer: str) -> (str, str):
+    def _parse_llm_response(self, answer: str) -> (str, str):
         """Parse next instruction of LLM output.
 
         Args:
             answer(str): output of LLM
 
         Returns:
-            Return a tuple, (action, action input)
+            Return a tuple, (thought,action, action input)
```
Comment (Undertone0809) [inline promptulate/agents/tool_agent/agent.py:171]: (thought, action, action input)  Pay attention to spaces

File: tests/hook/test_agent.py (hunk 10 above)
Comment (ruanrongman) [reply tests/hook/test_agent.py:91]: If this one doesn't have this key-value pair, then an error should be reported

File: promptulate/agents/tool_agent/agent.py (hunk 11 above)
Comment (ruanrongman) [reply promptulate/agents/tool_agent/agent.py:171]: ok

File: tests/hook/test_agent.py (hunk 10 above)
Comment (Undertone0809) [reply tests/hook/test_agent.py:90]: Therefore, if an error occurs, assertIsNotNone (line 91-93) is meaningless

**Review by Undertone0809 — APPROVED** [review]

**Pull request merged**